keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,423 @@
+  Azure/AWS do not configure their HCs to point to the service proxy. Instead
+  they connect to the `NodePort` defined for the service. As to have them
+  benefit from the proposals of this KEP, they would need to change their
+  implementation.
+
+## Proposal
+
+#### Risk
+
+The risk with each proposal is:
+
+1. By not being able to watch the Node object (while failing to read from the
+   API server, for example) we might have all Kube-proxy start failing the HCs
+   at once. That being said: Kube-proxy currently watches the Node object and is
+   susceptible to this risk.
+
+2. If Kube-proxy is flapping state often between healthy and unhealthy then
+   there's risk of disproportionate impact on `eTP:Local`, given that these
+   services won't have any other traffic option other than the node in question.
+
+3. Today: if Kube-proxy is unhealthy but an `eTP:Local` service has a ready
+   endpoint on the Node: then Kube-proxy will indicate to the load balancer that
+   it should send that node traffic. This will not be the case anymore once this
+   change is implemented. Clients relying on this behavior will probably see
+   their connections fail.
+
+#### Mitigations
+
+1. If Kube-proxy starts failing when reading from the API server, it should just
+   assume that the last state seen continues. For Kube-proxy to be made aware of
+   this, it needs to invoke
+   `serviceInformer.Informer().SetWatchErrorHandler(DefaultWatchErrorHandler)`
+   when initializing its informers. Any errors observed by client-go when
+   watching from the API server will be reported on `DefaultWatchErrorHandler`.
+
+2. Metrics should inform on Kube-proxy health and include information about its
+   `healthz` state, this can then be used to correlate to networking metrics
+   surrounding new/established connections on the node. Ex: a failing `healthz`
+   should correlate to a total drop in the count of new connections and with a
+   zero-or-negative rate of established connections. E2E tests should also be
+   designed with this specific goal in mind, i.e: validating the impact of a
+   failing kube-proxy on ingress connectivity. Kube-proxy currently has a lot of
+   metrics regarding how its health is doing, but no direct red/green indicator
+   of what the end result of its health is. One such metric could be
+   `proxy_health_check_200_count` / `proxy_health_check_503_count`
+
+3. The feature could be disabled for user who is dependent upon such behavior by
+   means of flipping the feature flag to off.
+
+## Design Details
+
+1. Implement change in Kube-proxy which allows the healthcheck server access to
+   the `healthz` state.
+
+2. Implement Kube-proxy change invoking client-go's `SetWatchErrorHandler`  on
+   watch errors from the API server.
+
+3. Implement change in Kube-proxy which will react to changes on the Node object
+   and once the taint `ToBeDeletedByClusterAutoscaler` is placed on the Node
+   object: start failing it's `healthz` state.
```

Comment:
Do I get it right, that it will influence both etp=Cluster and etp=Local health checks? (as etp=Local will report healthz status)

Reply:
Yeah, this part was left in limbo for a couple of hours until I got more opinions on: https://github.com/kubernetes/enhancements/pull/3837#discussion_r1100015632. It seems that the consensus gravitates towards removing the proposed changes for `eTP:Local` and just dealing with those issues as a bug. So, I will update the KEP and keep the proposed changes for `eTP:Cluster`

For reference, we don't want `ToBeDeletedByClusterAutoscaler` (or any other node indicator) to impact `eTP:Local` - but we do want to Kube-proxy's `healthz` to do so.

Reply:
`For reference, we don't want ToBeDeletedByClusterAutoscaler (or any other node indicator) to impact eTP:Local - but we do want Kube-proxy's healthz to do so.`

Why is so? Are there any problems with having possibility of the same draining, on `ToBeDeletedByClusterAutoscaler`  for `etp=Local`?

Reply:
Because of what we've discussed here: https://github.com/kubernetes/enhancements/pull/3837#discussion_r1099244706

Reply:
We discussed unschedulable there, but some concerns can be raised for `ToBeDeletedByClusterAutoscaler` too, like what if the CA places that on a Node but is unable to evict workloads that require ingress? That would cause Kube-proxy to block ingress for these workloads until this resolves.

Reply:
Suppose you have a single-endpoint service, running on node A, with eTP:Local.  Node A's kube-proxy passes for this service.  LBs route to node A, which routes to the local pod. All good.

Now we do a cluster node upgrade.  Eventually we get to node A.  We cordon it.  What should happen?

a) LBs stop sending traffic to node A because the HC starts failing
b) LBs continue sending traffic to node A until the pod itself becomes unready (presumably replaced by a new pod on a new node)

I think (b)

Reply:
I agree with b).

Reply:
agree with Alexander, b) ... _local Pod wins local HC_ :)